src/binary/level_4_entries.rs

```diff
@@ -13,35 +19,86 @@ pub struct UsedLevel4Entries {
         let mut used = UsedLevel4Entries {
             entry_state: [false; 512],
         };
 
         used.entry_state[0] = true; // TODO: Can we do this dynamically?
 
-        for segment in segments {
-            let start_page: Page = Page::containing_address(VirtAddr::new(
-                segment.virtual_addr() + virtual_address_offset,
-            ));
-            let end_page: Page = Page::containing_address(VirtAddr::new(
-                segment.virtual_addr() + virtual_address_offset + segment.mem_size(),
-            ));
+        // Mark the statically configured ranges from the config as used.
+
+        if let Some(physical_memory_offset) = CONFIG.physical_memory_offset {
+            used.mark_range_as_used(physical_memory_offset, max_phys_addr.as_u64().into_usize());
+        }
+
+        if CONFIG.map_page_table_recursively {
+            if let Some(recursive_index) = CONFIG.recursive_index {
+                used.mark_p4_index_as_used(PageTableIndex::new(recursive_index));
+            }
+        }
+
+        if let Some(kernel_stack_address) = CONFIG.kernel_stack_address {
+            used.mark_range_as_used(kernel_stack_address, CONFIG.kernel_stack_size());
+        }
+
+        if let Some(boot_info_address) = CONFIG.boot_info_address {
+            let boot_info_layout = Layout::new::<BootInfo>();
+            let regions = regions_len + 1; // one region might be split into used/unused
+            let memory_regions_layout = Layout::array::<MemoryRegion>(regions).unwrap();
+            let (combined, _) = boot_info_layout.extend(memory_regions_layout).unwrap();
+
+            used.mark_range_as_used(boot_info_address, combined.size());
+        }
 
-            for p4_index in u64::from(start_page.p4_index())..=u64::from(end_page.p4_index()) {
-                used.entry_state[p4_index as usize] = true;
+        if CONFIG.map_framebuffer {
+            if let Some(framebuffer_address) = CONFIG.framebuffer_address {
+                used.mark_range_as_used(framebuffer_address, framebuffer_size);
             }
         }
 
         used
     }
 
+    /// Mark all p4 entries in the range `[address..address+size)` as used.
+    ///
+    /// `size` can be a `u64` or `usize`.
+    fn mark_range_as_used<S>(&mut self, address: u64, size: S)
+    where
+        VirtAddr: core::ops::Add<S, Output = VirtAddr>,
+    {
+        let start = VirtAddr::new(address);
+        let end_inclusive = (start + size) - 1usize;
+        let start_page = Page::<Size4KiB>::containing_address(start);
+        let end_page_inclusive = Page::<Size4KiB>::containing_address(end_inclusive);
+
+        for p4_index in u16::from(start_page.p4_index())..=u16::from(end_page_inclusive.p4_index())
+        {
+            self.mark_p4_index_as_used(PageTableIndex::new(p4_index));
+        }
+    }
+
+    fn mark_p4_index_as_used(&mut self, p4_index: PageTableIndex) {
+        self.entry_state[usize::from(p4_index)] = true;
```

Comment:
Perhaps we should add an assertion here to prevent reusing the same p4 index?

Reply:
The problem with that is that we still want to support mapping multiple statically configured address ranges that share the same p4 index. I think if we want to check if statically configured address ranges overlap, we should do it somewhere else because we already lost too much information about the address here.

Alternatively we could replace `UsedLevel4Entries` entirely by something else that tracks the complete addresses not just the p4 indices. Given that we don't map that many addresses, we could probably just use a `StackVec` or something similar.

Reply:
Ah true! I agree that checking this somewhere else is probably a better solution. But we can do this in a separate PR.